OsmAnd/src/net/osmand/plus/server/ApiRouter.java

```diff
@@ -33,21 +30,95 @@ public OsmandApplication getAndroidContext() {
 	private final String FOLDER_NAME = "server";
 	private Gson gson = new Gson();
 	private Map<String, ApiEndpoint> endpoints = new HashMap<>();
+	//change to weakreference
+	public static MapActivity mapActivity;
 
-	public ApiRouter(){
-		initFavorites();
+	public ApiRouter() {
+		initRoutes();
 	}
 
-	private void initFavorites() {
+	private void initRoutes() {
 		ApiEndpoint favorites = new ApiEndpoint();
 		favorites.uri = "/favorites";
-		favorites.apiCall = new ApiEndpoint.ApiCall(){
+		favorites.apiCall = new ApiEndpoint.ApiCall() {
 			@Override
 			public NanoHTTPD.Response call(NanoHTTPD.IHTTPSession session) {
 				return newFixedLengthResponse(getFavoritesJson());
 			}
 		};
-		endpoints.put(favorites.uri,favorites);
+		endpoints.put(favorites.uri, favorites);
+
+		final ApiEndpoint tile = new ApiEndpoint();
+		tile.uri = "/tile";
+		tile.apiCall = new ApiEndpoint.ApiCall() {
+			@Override
+			public NanoHTTPD.Response call(NanoHTTPD.IHTTPSession session) {
+				try {
+					return tileApiCall(session);
+				} catch (Exception e) {
+					e.printStackTrace();
+				}
+				return ErrorResponses.response500;
+			}
+		};
+		endpoints.put(tile.uri, tile);
+	}
+
+	ExecutorService executor = Executors.newFixedThreadPool(3);
+
+	Map<RotatedTileBox, Bitmap> hashMap = new HashMap<>();
+	Map<RotatedTileBox, Bitmap> map = Collections.synchronizedMap(hashMap);
+
+	private synchronized NanoHTTPD.Response tileApiCall(NanoHTTPD.IHTTPSession session) {
```

Comment:
bad name - method synchronized should be only for rendering tile, all parsing lat, lon, zoom should be parent method and not synchronized

requestTile(String url)

synchronized renderTile (int x, int y, int z)